src/PureScript/Backend/Optimizer/Convert.purs

```diff
@@ -264,17 +264,24 @@ inferTransitiveDirective directives impl backendExpr cfn = fromImpl <|> fromBack
       case Map.lookup (EvalExtern qual) directives of
         Just dirs -> do
           let
-            newDirs = Map.toUnfoldable dirs # Array.mapMaybe case _ of
-              Tuple InlineRef (InlineArity n) ->
-                Just $ Tuple InlineRef (InlineArity (n - NonEmptyArray.length args))
-              Tuple (InlineSpineProp prop) dir ->
-                Just $ Tuple (InlineProp prop) dir
-              _ ->
-                Nothing
-          if Array.null newDirs then
+            newDirs = foldrWithIndex
+              ( \ix dir accum -> case ix, dir of
+                  InlineRef, (InlineArity n) ->
+                    accum
+                      # Map.insert InlineRef (InlineArity (n - NonEmptyArray.length args))
+                  InlineSpineProp prop, _ ->
+                    accum
+                      # Map.insert (InlineProp prop) dir
+                      # Map.insert (InlineSpineProp prop) dir
```

Comment:
We still need to keep the old one around because the application spine may be split up among multiple bindings.